src/components/application_manager/src/commands/command_impl.cc

```diff
@@ -94,14 +94,59 @@ void CommandImpl::SetAllowedToTerminate(const bool allowed) {
   allowed_to_terminate_ = allowed;
 }
 
+bool CommandImpl::ReplaceMobileWithHMIAppId(
+    NsSmartDeviceLink::NsSmartObjects::SmartObject& message) {
+  if (message.keyExists(strings::app_id)) {
+    ApplicationSharedPtr application =
+        application_manager_.application(message[strings::app_id].asUInt());
+    if (!application) {
+      LOG4CXX_ERROR(logger_, "Substitution mobile --> HMI id is failed.");
+      return false;
+    }
+    LOG4CXX_DEBUG(logger_,
+                  "ReplaceMobileWithHMIAppId from "
+                      << message[strings::app_id].asInt() << " to "
+                      << application->hmi_app_id());
+    message[strings::app_id] = application->hmi_app_id();
+  } else {
+    switch (message.getType()) {
+      case smart_objects::SmartType::SmartType_Array: {
+        smart_objects::SmartArray* message_array = message.asArray();
+        smart_objects::SmartArray::iterator it = message_array->begin();
+        for (; it != message_array->end(); ++it) {
+          if (!ReplaceMobileWithHMIAppId(*it)) {
+            break;
```

Comment:
break will cause the function return true, however it shall return false here

Reply:
@yang1070 
Done with 866dfe8

Reply:
@yang1070 
Reverted previous changes due to issues with ATF tests. As I found there is a mixed logic for two use cases: first is related to bugfix done for Ford Sync transport switch implementation which requires checking of application with app_id, second is common usage when app_id is replaced during message processing. So I propose to leave code as is for now and create another issue for investigation how this part can be re-factored.

Reply:
break and return true is worse than not checking the return value

Reply:
@yang1070 
Reverted changes are breaking changes. From other side I w/o that everything works as expected. So that's why I propose to put this into separate issue.